src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PETypeParameterSymbol.cs

```diff
@@ -234,31 +254,44 @@ public override bool HasConstructorConstraint
         {
             get
             {
-                return (_flags & GenericParameterAttributes.DefaultConstructorConstraint) != 0;
+                GetDeclaredConstraintTypes();
+                return (_lazyFlags & GenericParameterAttributes.DefaultConstructorConstraint) != 0;
             }
         }
 
         public override bool HasReferenceTypeConstraint
         {
             get
             {
-                return (_flags & GenericParameterAttributes.ReferenceTypeConstraint) != 0;
+                GetDeclaredConstraintTypes();
+                return (_lazyFlags & GenericParameterAttributes.ReferenceTypeConstraint) != 0;
             }
         }
 
         public override bool HasValueTypeConstraint
         {
             get
             {
-                return (_flags & GenericParameterAttributes.NotNullableValueTypeConstraint) != 0;
+                GetDeclaredConstraintTypes();
+                return (_lazyFlags & GenericParameterAttributes.NotNullableValueTypeConstraint) != 0 || HasUnmanagedTypeConstraint;
+            }
+        }
+
+        public override bool HasUnmanagedTypeConstraint
+        {
+            get
+            {
+                GetDeclaredConstraintTypes();
```

Comment:
>GetDeclaredConstraintTypes(); [](start = 16, length = 29)

I don't think this change is necessary , presence of an attribute can be checked independently as we do everywhere else in PE symbols.

Reply:
On the other hand, driving result of off an attribute is probably not the right thing, the custom modifier on ValueType is probably a better indicator.

---
In reply to: [172413190](https://github.com/dotnet/roslyn/pull/24817#discussion_r172413190) [](ancestors = 172413190)